export/mlab_export.py

```diff
@@ -0,0 +1,605 @@
+  Raises:
+    IOError, if last_export_filename does not exist and cannot be created.
+    OSError, if last_export_filename exists but stat info cannot be read.
+  """
+  if not os.path.exists(last_export_filename):
+    open(last_export_filename, 'w').close()
+    # Indicate that there is no timestamp, so the caller can use a default.
+    return 0
+  return int(os.stat(last_export_filename).st_mtime)
+
+
+def update_mtime(last_export_filename, mtime):
+  """Updates the atime & mtime on the export timestamp file.
+
+  Args:
+    last_export_filename: str, absolute path to last export time stamp file.
+    mtime: int, timestamp in seconds since epoch; used for file atime & mtime.
+  Raises:
+    OSError, if last_export_filename mtime cannot be updated.
+  """
+  os.utime(last_export_filename, (mtime, mtime))
+
+
+def align_timestamp(timestamp, step):
+  """Adjusts 'timestamp' down to a multiple of 'step' size.
+
+  Args:
+    timestamp: int, timestamp in seconds since the epoch.
+    step: int, interval to align.
+  Returns:
+    int, timestamp adjusted to multiple of step.
+  """
+  return timestamp - (timestamp % step)
+
+
+def default_start_time(options, ts_previous):
+  """Calculates a default start timestamp.
+
+  Args:
+    options: flags.FlagValues, the runtime options. These values are read:
+        options.length, options.step, options.update, options.ts_offset.
+    ts_previous: int, timestamp in seconds since epoch of last
+        successful export. On first run, this value should be zero.
+  Returns:
+    int, timestamp in seconds since the epoch.
+  """
+  if ts_previous:
+    # Typical: start from the end time of previous runs.
+    return align_timestamp(ts_previous, options.step)
+
+  ts_current = int(time.time())
+
+  # Since ts_previous is not set, this is a "first run" scenario.
+  if not options.update:
+    # Unlikely: first run by a user. Start as close to 'now' as possible.
+    start = ts_current - options.length - options.ts_offset
+  else:
+    # Likely: first, automated export. Start at previous 'length' aligned time.
+    start = align_timestamp(ts_current, options.length) - options.length
+
+  # Align end ts to a multiple of step size (just in case).
+  return align_timestamp(start, options.step)
+
+
+def assert_start_and_end_times(options):
+  """Performs a sanity check on start and end timestamps.
+
+  This method asserts that both ts_end is less than ts_start and that the
+  difference between them is greater than options.length.
+
+  Args:
+    options: flags.FlagValues, the runtime options. These values are read:
+        options.length, options.ts_start, options.ts_end.
+  Returns:
+    True on success.
+  Raises:
+    TimeOptionError, if a start & end time constraint is violated.
+  """
+  # Always check if basic constratins are respected.
+  if options.ts_end <= options.ts_start:
+    raise TimeOptionError('Start time must precede end time.')
+
+  if options.ts_end - options.ts_start != options.length:
```

Comment:
Good question & point. Some of the optional flags would cause this check to fail if given explicitly with miss-aligned values.

My hope with the flags was mostly to make more manual diagnostics by operators possible than if the values were static or unexposed, but I didn't image they would be guaranteed to work for all situations. I've added clarifying text to the `--length` and `--ts_start` flags that length should be a multiple of step size, and ts_start will normally be automatic.

Is this an okay trade-off?

e7d0bc1277fc48285ee5aece16c23f37356edca4...fdefcd11b9f0adb117972ed77ac13352c4cd0532